### PR TITLE
Add numerical derivative to Signal

### PR DESCRIPTION
**Pull request by francisco-dlp**
Example of usage:

```python

>>> import hyperspy.hspy as hs
>>> import numpy as np
>>> offset = 3
>>> scale = 0.01
>>> x = np.arange(-offset, offset, scale)
>>> s = signals.Spectrum(np.sin(x))
>>> s.axes_manager[0].scale = scale
>>> s.axes_manager[0].offset = -offset
>>> utils.plot.plot_spectra([s.derivative(0, i) for i in range(5)])
<matplotlib.axes._subplots.AxesSubplot object at 0x7f09d0d0acd0>


```
derivative

### CODE REVIEW

**Comment by francisco-dlp** [comment]
Note that, although this is a new feature, I'm pull requesting it to 0.8.x because it is required for #588.

File: hyperspy/signal.py [hunk 1]
```diff
@@ -3810,16 +3807,46 @@ def diff(self, axis, order=1):
...
-        axis.offset += (axis.scale / 2)
+            np.diff(self.data,
+                    n=order,
+                    axis=self.axes_manager[axis].index_in_array))
+        axis = s.axes_manager[axis]
+        axis.offset += (order * axis.scale / 2)
         s.get_dimensions_from_data()
         return s
 
+    def derivative(self, axis, order=1, central_differences=False):
```
Comment (to266) [inline hyperspy/signal.py]: `central_differences` is not used (and not clear what it does as well)

Comment (francisco-dlp) [inline hyperspy/signal.py]: Fixed, thanks!

File: hyperspy/tests/signal/test_tools.py [hunk 2]
```diff
@@ -318,3 +331,20 @@ def test_signal_iterator():
...
+        offset = 3
+        scale = 0.1
+        x = np.arange(-offset, offset, scale)
+        s = signals.Spectrum(np.sin(x))
+        s.axes_manager[0].offset = x[0]
+        s.axes_manager[0].scale = scale
+        self.s = s
+
+    def test_derivative_data(self):
+        self.s.derivative(axis=0, order=4)
```
Comment (to266) [inline hyperspy/tests/signal/test_tools.py]: Here you don't save the derivative signal, hence don't test it in any way...

File: hyperspy/signal.py [hunk 3]
```diff
@@ -3808,16 +3805,46 @@ def diff(self, axis, order=1):
...
+        order: int
+            The order of the derivative
+
+        Returns
+        -------
+        def : Signal
+            The size of the
+
+        See also
+        --------
```
Comment (to266) [inline hyperspy/signal.py]: I would have a look at the documentation again, to make sure that it states correct things and remove any unfinished sentences.

Comment (francisco-dlp) [inline hyperspy/signal.py]: It should be fixed now, thanks again!

**Pull request merged**